### PR TITLE
Tbl

### PR DESCRIPTION
**Pull request by sigfredonin**
This is still half-baked. 
Finished parts 1 and 2 of 4 for problem 5.6.
1. Divide corpus sentences 80:20 into training and test tests. Verify test set has words not in training set.
2. For training set, find most frequent tag for each word.
3. For the words in the test set …
   * count errors if tagged with 'NN' (all assumed to be nouns)
   * count errors if tagged with most frequent tag for the word
   Note: compute separate errors for known and unknown words.
4. Create 5 templates for the Brill TBL algorithm and tag the test set.
   * count errors, again for known and unknown words separately.

### CODE REVIEW

**Comment by sigfredonin** [comment]
Part 3 of problem 5.6 completed.

Improvement from blind tagging as nouns to tagging with most frequent tag evident.
When all words tagged nouns, no significant difference in error rate between known and unknown
words - I think the corpus is too small for the training set to capture enough of the vocabulary
so that new words in the test set are mostly names.

**Comment by sigfredonin** [comment]
Problem 5.5 Implement the Brill TBL algorithm in Figure 5.21.  Create a small number of templates and train the tagger on any POS-tagged training set you can find.
1. The algorithm is implemented, in class BrillTBL, and has been tested with training sets from the NLKT treebank corpus, tagged with the 45-tag tagset.  Each test seems to find 1 or 2 rules of the form
"Replace [fromtag] with [toTag] iff the preceding word is tagged with [zTag]".

Still to come:
2. Implement the "Apply Rules" algorithm to tag a corpus with the learned rule sequence.
3. Create the additional templates needed to complete problem 5.6.

**Comment by sigfredonin** [comment]
After further refinement, testing the implementation was taking over two hours to run the learn algorithm.  This makes testing very burdensome, so implemented a "data-driven" strategy: to test a template, iterate over the corpus, and for each tagged word, use the word's tag as the fromTag, the word's true tag as the correctTag, and the previous word's tag as the zTag; iterate over the tags in the tagset as the toTag to produces transforms [ fromTag, toTag, zTag ] and evaluate them.

This change reduces the time to run the learn algorithm to about 2-3/4 minutes.
